src/go/rpk/pkg/cli/cluster/partitions/move.go

```diff
@@ -0,0 +1,327 @@
+			g, egCtx := errgroup.WithContext(cmd.Context())
+			for _, partition := range partitions {
+				partition := partition
+				g.Go(func() error {
+					if len(topics) > 0 { // foo -p 0:1,2,3
+						for _, t := range topics {
+							_, _, part, err := extractNTP(t, partition)
+							out.MaybeDie(err, "failed to extract topic/partition: %s\n", err)
+
+							if nt := strings.Split(t, "/"); len(nt) == 1 {
+								ns = "kafka"
+								topic = nt[0]
+							} else {
+								ns = nt[0]
+								topic = nt[1]
+							}
+
+							current, err := cl.GetPartition(egCtx, ns, topic, part)
+							out.MaybeDie(err, "unable to get partition: %s\n", err)
+
+							newReplicas, err := configureReplicas(partition, current.Replicas)
+							out.MaybeDie(err, "unable to configure new replicas: %v\n", err)
+
+							mu.Lock()
+							newAssignmentList = append(newAssignmentList, newAssignment{
+								Namespace:   ns,
+								Topic:       topic,
+								Partition:   part,
+								OldReplicas: current.Replicas,
+								NewReplicas: newReplicas,
+							})
+							mu.Unlock()
+						}
+					} else { // -p foo/0:1,2,3
+						ns, topic, part, err := extractNTP("", partition)
+						out.MaybeDie(err, "failed to extract topic/partition: %s\n", err)
+
+						current, err := cl.GetPartition(egCtx, ns, topic, part)
+						out.MaybeDie(err, "unable to get partition: %s\n", err)
+
+						newReplicas, err := configureReplicas(partition, current.Replicas)
+						out.MaybeDie(err, "unable to configure new replicas: %v\n", err)
+
+						mu.Lock()
+						newAssignmentList = append(newAssignmentList, newAssignment{
+							Namespace:   ns,
+							Topic:       topic,
+							Partition:   part,
+							OldReplicas: current.Replicas,
+							NewReplicas: newReplicas,
+						})
+						mu.Unlock()
+					}
+					return nil
+				})
+			}
+
+			if err := g.Wait(); err != nil {
+				out.Die("failed to parse the arguments: %v\n", err)
+			}
+
+			for _, newa := range newAssignmentList {
+				for _, nr := range newa.NewReplicas {
+					if nr.Core == -1 {
+						brokerReqs[nr.NodeID] = struct{}{}
+					}
+				}
+			}
+			for node := range brokerReqs {
+				node := node
+				g.Go(func() error {
+					broker, err := cl.Broker(cmd.Context(), node)
+					mu.Lock()
+					defer mu.Unlock()
+					knownNodeCore[node] = broker.NumCores
+					return err
+				})
+			}
+
+			if err := g.Wait(); err != nil {
+				out.Die("unable to find core counts", err)
+			}
+
+			rng := rand.New(rand.NewSource(time.Now().UnixNano()))
+			for i, newa := range newAssignmentList {
+				i := i
+				newa := newa
+				for j, nr := range newa.NewReplicas {
+					if nr.Core == -1 {
+						numCore := knownNodeCore[nr.NodeID]
+						newa.NewReplicas[j].Core = rng.Intn(numCore)
+					}
+				}
+				wg.Add(1)
+				go func(newa newAssignment) {
+					defer wg.Done()
+					err := cl.MoveReplicas(cmd.Context(), newa.Namespace, newa.Topic, newa.Partition, newa.NewReplicas)
+					mu.Lock()
+					defer mu.Unlock()
```

Comment:
fwiw this mutex isn't necessary because newAssignmentList[i] cannot conflict with a write by a different goroutine (since `i` will be different)